akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -754,8 +754,7 @@ import akka.util.OptionVal
   }
 
   private def shortCircuitBatch(): Unit = {
-    while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && activeInterpreters.nonEmpty) shortCircuitBuffer
-      .poll() match {
+    while (!shortCircuitBuffer.isEmpty && currentLimit > 0) shortCircuitBuffer.poll() match {
```

Comment:
:+1: this strictly runs more times than before which is good. If it runs even if `activeInterpreters.isEmpty`, `processEvent` below will just skip event processing because `shell.isTerminated` will be false.

Reply:
```suggestion
    @tailrec private def shortCircuitBatch(): Unit = {
    if(shortCircuitBuffer.isEmpty) ()
    else if (currentLimit == 0) {
      self ! Resume
    }
    else {
      shortCircuitBuffer.poll() match {
        case b: BoundaryEvent => processEvent(b)
        case Resume           => finishShellRegistration()
      }
      shortCircuitBatch()
    }
  }
```

@jrudolph , how do u feel about making a `tailrec` out of it?

Reply:
Sounds good to me :)

Reply:
Ah, that suggestion didn't replace the complete old method but only one line of it. Can you fix that?